bridges/snowbridge/parachain/primitives/router/src/inbound/mod.rs

```diff
@@ -0,0 +1,334 @@
+				0,
+			),
+			// Final destination is a 32-byte account on a sibling of AssetHub
+			Destination::ForeignAccountId32 { para_id, id, fee } => (
+				Some(para_id),
+				MultiLocation { parents: 0, interior: X1(AccountId32 { network: None, id }) },
+				// Total fee needs to cover execution on AssetHub and Sibling
+				fee,
+			),
+			// Final destination is a 20-byte account on a sibling of AssetHub
+			Destination::ForeignAccountId20 { para_id, id, fee } => (
+				Some(para_id),
+				MultiLocation { parents: 0, interior: X1(AccountKey20 { network: None, key: id }) },
+				// Total fee needs to cover execution on AssetHub and Sibling
+				fee,
+			),
+		};
+
+		let total_fees = asset_hub_fee.saturating_add(dest_para_fee);
+		let total_fee_asset: MultiAsset = (MultiLocation::parent(), total_fees).into();
+		let inbound_queue_pallet_index = InboundQueuePalletInstance::get();
+
+		let mut instructions = vec![
+			ReceiveTeleportedAsset(total_fee_asset.into()),
+			BuyExecution { fees: asset_hub_fee_asset, weight_limit: Unlimited },
+			DescendOrigin(X1(PalletInstance(inbound_queue_pallet_index))),
+			UniversalOrigin(GlobalConsensus(network)),
+			ReserveAssetDeposited(asset.clone().into()),
+			ClearOrigin,
+		];
+
+		match dest_para_id {
+			Some(dest_para_id) => {
+				let dest_para_fee_asset: MultiAsset =
+					(MultiLocation::parent(), dest_para_fee).into();
+
+				instructions.extend(vec![
+					// Perform a deposit reserve to send to destination chain.
+					DepositReserveAsset {
+						assets: Definite(vec![dest_para_fee_asset.clone(), asset.clone()].into()),
+						dest: MultiLocation { parents: 1, interior: X1(Parachain(dest_para_id)) },
+						xcm: vec![
+							// Buy execution on target.
+							BuyExecution { fees: dest_para_fee_asset, weight_limit: Unlimited },
+							// Deposit asset to beneficiary.
+							DepositAsset { assets: Definite(asset.into()), beneficiary },
+						]
+						.into(),
+					},
+				]);
+			},
+			None => {
+				instructions.extend(vec![
+					// Deposit asset to beneficiary.
+					DepositAsset { assets: Definite(asset.into()), beneficiary },
+				]);
+			},
+		}
+
+		(instructions.into(), total_fees.into())
+	}
+
+	// Convert ERC20 token address to a Multilocation that can be understood by Assets Hub.
+	fn convert_token_address(network: NetworkId, token: H160) -> MultiLocation {
+		MultiLocation {
+			parents: 2,
+			interior: X2(
+				GlobalConsensus(network),
+				AccountKey20 { network: None, key: token.into() },
+			),
+		}
+	}
+}
+
+pub struct FromEthereumGlobalConsensus<EthereumBridgeLocation>(PhantomData<EthereumBridgeLocation>);
+impl<EthereumBridgeLocation> ContainsPair<MultiLocation, MultiLocation>
+	for FromEthereumGlobalConsensus<EthereumBridgeLocation>
+where
+	EthereumBridgeLocation: Get<MultiLocation>,
+{
+	fn contains(asset: &MultiLocation, origin: &MultiLocation) -> bool {
+		origin == &EthereumBridgeLocation::get() && asset.starts_with(origin)
+	}
+}
```

Comment:
I am not sure about this, there is not doc, and  it does nothing specific  to `Ethereum` or `GlobalConsensus` as name `FromEthereumGlobalConsensus` suggests.
And for `ForeignCreators` where it is only used, it can be totally replaced with `FromNetwork` struct.

```suggestion
```